src/util/leveling.rs

```diff
@@ -0,0 +1,94 @@
+use chrono::prelude::*;
+use serenity::{Result, framework::standard::CommandResult};
+use std::sync::Arc;
+
+use redis::Commands;
+
+use tracing::{error, info, instrument};
+
+#[derive(Clone, Debug)]
+pub struct LevelData {
+    pub msg_count: u32,
+    pub xp: u32,
+    pub level: u32,
+    pub last_msg: chrono::DateTime<Utc>,
+}
+
+fn get_level_number(xp: u32) -> u32 {
+    let xp = xp as f64;
+    let mut level = (xp / 50f64).powf(1f64/3f64).floor();
+
+    if level < 0f64 {
+        level = 0f64;
+    }
+
+    level as u32
+}
+
+fn get_level_cost(level: u64) -> u64 {
+    (level.pow(3) * 50).into()
+}
+
+pub fn get_user_level(user_id: u64, redis_conn: &mut redis::Connection) -> Result<LevelData> {
+    let msg_count = match redis::cmd("GET").arg(&[format!("{}:count", &user_id)]).query(redis_conn) {
+        Ok(count) => count,
+        _ => 0
+    }; 
+    let xp: u32 = match redis::cmd("GET").arg(&[format!("{}:exp", &user_id)]).query(redis_conn) {
+        Ok(xp) => xp,
+        _ => 0,
+    };
+    let last_msg_num: i64 = match redis::cmd("GET").arg(&[format!("{}:last", &user_id)]).query(redis_conn) {
+        Ok(msg) => msg,
+        _ => 0,
+    };
+
+    let last_msg = Utc.timestamp(last_msg_num, 0);
+    let level = get_level_number(xp.into());
+
+    let level_data = LevelData {
+        msg_count,
+        xp,
+        last_msg,
+        level,
+    };
+        
+    Ok(level_data)
+}
+
+#[instrument(skip(redis_conn))]
+pub fn set_user_level(user_id: u64, redis_conn: &mut redis::Connection, level_data: LevelData) -> Result<()> {
+    redis_conn.set::<String, u32, ()>(format!("{}:count", user_id), level_data.msg_count);
+    redis::cmd("SET").arg(&[format!("{}:exp", &user_id), level_data.xp.to_string()]).query::<()>(redis_conn);
+    redis::cmd("SET").arg(&[format!("{}:last", &user_id), level_data.last_msg.timestamp().to_string()]).query::<()>(redis_conn);
```

Comment:
These are using two different ways to achieve the same thing? Is this necessary?

Reply:
No, it's left over from our testing.